typeurl/types.go

```diff
@@ -0,0 +1,126 @@
+package typeurl
+
+import (
+	"encoding/json"
+	"path"
+	"reflect"
+	"strings"
+	"sync"
+
+	"github.com/containerd/containerd/errdefs"
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/types"
+)
+
+const Prefix = "types.containerd.io"
+
+var (
+	mu       sync.Mutex
+	registry = make(map[reflect.Type]string)
+)
+
+// Register a type with the base url of the type
+func Register(v interface{}, args ...string) {
+	t := tryDereference(v)
+	mu.Lock()
+	defer mu.Unlock()
+	if _, ok := registry[t]; ok {
+		panic(errdefs.ErrAlreadyExists)
+	}
+	registry[t] = path.Join(append([]string{Prefix}, args...)...)
+}
+
+// TypeURL returns the type url for a registred type
+func TypeURL(v interface{}) (string, error) {
+	mu.Lock()
+	defer mu.Unlock()
+	u, ok := registry[tryDereference(v)]
+	if !ok {
+		// fallback to the proto registry if it is a proto message
+		pb, ok := v.(proto.Message)
+		if !ok {
+			return "", errdefs.ErrNotFound
+		}
+		return path.Join(Prefix, proto.MessageName(pb)), nil
+	}
+	return u, nil
+}
+
+func Is(any *types.Any, v interface{}) bool {
+	url, err := TypeURL(v)
+	if err != nil {
+		return false
+	}
+	return any.TypeUrl == url
+}
+
+func MarshalAny(v interface{}) (*types.Any, error) {
+	var (
+		err  error
+		data []byte
+	)
+	url, err := TypeURL(v)
+	if err != nil {
+		return nil, err
+	}
+	switch t := v.(type) {
+	case proto.Message:
+		data, err = proto.Marshal(t)
+	default:
+		data, err = json.Marshal(v)
+	}
+	if err != nil {
+		return nil, err
+	}
+	return &types.Any{
+		TypeUrl: url,
+		Value:   data,
+	}, nil
+}
+
+func UnmarshalAny(any *types.Any) (interface{}, error) {
+	t, err := getTypeByUrl(any.TypeUrl)
+	if err != nil {
+		return nil, err
+	}
+	v := reflect.New(t.t).Interface()
```

Comment:
Inside of `EmptyAny`, they always do `reflect.New(t.t.Elem()).Interface()`, which indicates they always work with pointer types for registration. We can enforce a pointer type at registration to prevent the wrong type from being registered.